File: PGBooks/PGBooks/Sources/UI/Common/Detail/BookDetailController.m
//
//  BookDetailController.m
//  PGBooks
//
//  Created by ipagong on 13/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

#import "BookDetailController.h"
#import "CustomViews.h"
#import "PGBook.h"
#import "ITBookService.h"
#import "UserDataService.h"
#import "UIImageView+AFNetworking.h"
#import "NSString+Utils.h"
#import <PGTransitionKit-Swift.h>

@interface BookDetailController ()
@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;

@property (weak, nonatomic) IBOutlet UIImageView *coverView;

@property (weak, nonatomic) IBOutlet UIStackView *stackContainer;

@property (weak, nonatomic) IBOutlet PaddingLabel *priceLabel;
@property (weak, nonatomic) IBOutlet UIButton *linkButton;
@property (weak, nonatomic) IBOutlet UIButton *favoriteButton;

@property (weak, nonatomic) IBOutlet UILabel *isbn13Label;
@property (weak, nonatomic) IBOutlet UILabel *isbn10Label;


@property (weak, nonatomic) IBOutlet UIStackView *ratingContainer;

@property (weak, nonatomic) IBOutlet UILabel *bookTitleLabel;
@property (weak, nonatomic) IBOutlet UILabel *bookSubtitleLabel;
@property (weak, nonatomic) IBOutlet UILabel *etcLabel;

@property (weak, nonatomic) IBOutlet UILabel *authorsLabel;
@property (weak, nonatomic) IBOutlet UILabel *publisherLabel;
@property (weak, nonatomic) IBOutlet UILabel *languagesLabel;

@property (weak, nonatomic) IBOutlet UILabel *descLabel;

@property (nonatomic, strong) PGAnimator * animator;
@property (nonatomic, strong) PGScrollDownInteractor * interactor;

@end

@implementation BookDetailController

#pragma mark - create

+ (void)showWithBook:(PGBook *)book target:(UIViewController *)target {
    UIStoryboard * storyboard = [UIStoryboard storyboardWithName:@"BookDetail" bundle:nil];
    BookDetailController *vc = [storyboard instantiateViewControllerWithIdentifier:NSStringFromClass([BookDetailController class])];
    [UserDataService addGlanceBook:book];
    vc.book = book;
    
    vc.modalPresentationStyle = UIModalPresentationFullScreen;
    
    vc.animator = [[PGAnimator alloc] initWithTarget:target presenting:vc];
    vc.animator.presentDuration = 0.3;
    vc.animator.dismissDuration = 0.3;
    
    [vc.animator presentAnimatorViewControllerWithAnimated:YES];
}

#pragma mark - life cycle

- (void)viewDidLoad {
    [super viewDidLoad];
 
    [self setupUI];
    
    self.interactor = [[PGScrollDownInteractor alloc] initWithTargetView:self.scrollView];
    self.animator.dismissInteractor = self.interactor;
    
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    [self dataLoad];
}

#pragma mark -

- (void)setupUI {
    if (!self.book) { return; }
    
    [self.priceLabel        setText:[@"----" orEmpty:self.book.price]];
    [self.isbn13Label       setText:[@"----" orEmpty:self.book.isbn13]];
    [self.isbn10Label       setText:[@"----" orEmpty:self.book.isbn10]];
    [self.bookTitleLabel    setText:[@"----" orEmpty:self.book.title]];
    [self.bookSubtitleLabel setText:[@"----" orEmpty:self.book.subtitle]];
    [self.etcLabel          setText:[@"----" orEmpty:self.book.etcs]];
    [self.authorsLabel      setText:[@"----" orEmpty:self.book.authors]];
    [self.publisherLabel    setText:[@"----" orEmpty:self.book.publisher]];
    [self.languagesLabel    setText:[@"----" orEmpty:self.book.language]];
    [self.descLabel         setText:[@"----" orEmpty:self.book.desc]];
    
    [self.favoriteButton setSelected:[[UserDataService favoriteBooks] containsObject:self.book]];
    
    if (self.book.image) {
        [self.coverView setImageWithURL:[NSURL URLWithString:self.book.image]];
    }
    
    for (UIImageView *imgView in self.ratingContainer.subviews) {
        if ([imgView isKindOfClass:[UIImageView class]] == NO) { return; }
        if (!self.book.rating) { [imgView setHighlighted:false]; }
        BOOL on = ([self.book.rating integerValue] >= imgView.tag);
        [imgView setHighlighted:on];
    }
}

- (IBAction)onClickFavorite:(id)sender {
    if ([[UserDataService favoriteBooks] containsObject:self.book] == YES) {
        [UserDataService removeFavoriteBookWithObject:self.book];
    } else {
        [UserDataService addFavoriteBook:self.book];
    }
    
    [self setupUI];
}

- (IBAction)onClickLink:(id)sender {
    if (self.book.url == nil) { return; }
    
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:self.book.url]
                                       options:@{}
                             completionHandler:nil];
}

- (IBAction)onClickClose:(id)sender {
    [self.animator dismissAnimatorViewControllerWithAnimated:YES];
}

- (void)dataLoad {
    if (self.book.isbn10 && self.book.isbn10.length > 0) { return; }
    
    __weak __typeof(self)weakSelf = self;
    [ITBookService getDetailBookInfo:self.book.isbn13
                          completion:^(PGBook * _Nonnull book, NSError * _Nonnull error) {
                              __strong __typeof(self)strongSelf = weakSelf;
                              strongSelf.book = book;
                              [UIView animateWithDuration:0.3 animations:^{
                                  [weakSelf setupUI];
                                  [weakSelf.view setNeedsLayout];
                                  [weakSelf.view layoutIfNeeded];
                              }];
                          }];
}

@end

File: PGBooks-swift/PGBooks-swift/Sources/UI/Common/Detail/BookDetailController.swift
//
//  BookDetailController.swift
//  PGBooks-swift
//
//  Created by ipagong on 16/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

import UIKit
import RxSwift
import RxCocoa
import RxDataSources
import PGTransitionKit

protocol BookDetailTransitionable {
    var tableview:UITableView! { get }
}

class BookDetailController: UIViewController {
    @IBOutlet var scrollView        : UIScrollView!
    @IBOutlet var coverView         : UIImageView!
    @IBOutlet var stackContainer    : UIStackView!
    @IBOutlet var priceLabel        : PaddingLabel!
    @IBOutlet var linkButton        : UIButton!
    @IBOutlet var favoriteButton    : UIButton!
    @IBOutlet var isbn13Label       : UILabel!
    @IBOutlet var isbn10Label       : UILabel!
    @IBOutlet var ratingContainer   : UIStackView!
    @IBOutlet var bookTitleLabel    : UILabel!
    @IBOutlet var bookSubtitleLabel : UILabel!
    @IBOutlet var etcLabel          : UILabel!
    @IBOutlet var authorsLabel      : UILabel!
    @IBOutlet var publisherLabel    : UILabel!
    @IBOutlet var languagesLabel    : UILabel!
    @IBOutlet var descLabel         : UILabel!
    
    @IBOutlet weak var closeButton    : UIButton!
    
    var animator: PGTransitionKit.Animator!
    
    public var book:Book!
    
    private let viewModel = BookDetailViewModel.init()
    private let disposeBag = DisposeBag()
    
    static func createInstance() -> BookDetailController? {
        return UIStoryboard.init(name: "BookDetail", bundle: nil).instantiateViewController(withIdentifier: "BookDetailController") as? BookDetailController
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        binding()
    }

    func setupUI() {
        self.animator.dismissInteractor = ScrollDownInteractor.init(targetView: self.scrollView)
    }
    
    func binding() {
        self.viewModel.output.price     .bind(to: priceLabel.rx.text)          .disposed(by: self.disposeBag)
        self.viewModel.output.isbn13    .bind(to: isbn13Label.rx.text)         .disposed(by: self.disposeBag)
        self.viewModel.output.title     .bind(to: bookTitleLabel.rx.text)      .disposed(by: self.disposeBag)
        self.viewModel.output.subtitle  .bind(to: bookSubtitleLabel.rx.text)   .disposed(by: self.disposeBag)
        self.viewModel.output.isbn10    .bind(to: isbn10Label.rx.text)         .disposed(by: self.disposeBag)
        self.viewModel.output.etcs      .bind(to: etcLabel.rx.text)            .disposed(by: self.disposeBag)
        self.viewModel.output.authors   .bind(to: authorsLabel.rx.text)        .disposed(by: self.disposeBag)
        self.viewModel.output.publisher .bind(to: publisherLabel.rx.text)      .disposed(by: self.disposeBag)
        self.viewModel.output.language  .bind(to: languagesLabel.rx.text)      .disposed(by: self.disposeBag)
        self.viewModel.output.desc      .bind(to: descLabel.rx.text)           .disposed(by: self.disposeBag)
        self.viewModel.output.image     .bind(to: coverView.rx.image)          .disposed(by: self.disposeBag)
        self.viewModel.output.favorite  .bind(to: favoriteButton.rx.isSelected).disposed(by: self.disposeBag)
        
        self.viewModel.output.rating
            .subscribe(onNext: { [weak self](rating) in
                self?.ratingContainer.subviews
                    .compactMap{ $0 as? UIImageView }
                    .forEach{ $0.isHighlighted = ($0.tag <= rating)}
            }).disposed(by: self.disposeBag)
        
        self.lifeCycle.signalViewWillAppear()
            .map{ [weak self] in self?.book }
            .bind(to: viewModel.input.loadSignal)
            .disposed(by: self.disposeBag)
        
        self.linkButton.rx.tap
            .bind(to: self.viewModel.input.link)
            .disposed(by: self.disposeBag)
        
        self.favoriteButton.rx.tap
            .bind(to: self.viewModel.input.favoriteToggle)
            .disposed(by: self.disposeBag)
        
        self.closeButton.rx.tap
            .subscribe(onNext: { [weak self] in self?.animator.dismissAnimatorViewController() })
            .disposed(by: self.disposeBag)
        
        
    }
}

extension BookDetailTransitionable where Self : UIViewController {
    func showBookDetail(_ book:Book) {
        guard let vc = BookDetailController.createInstance() else { return }
        vc.book = book
        
        vc.modalPresentationStyle = .fullScreen
    
        vc.animator = Animator.init(target: self, presenting: vc)
        vc.animator.presentDuration = 0.3
        vc.animator.dismissDuration = 0.3

        vc.animator.presentAnimatorViewController()
    }
}
